### PR TITLE
Add parser for codechecker

### PR DESCRIPTION
**Pull request by ferkulat**
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

Codechecker is a wrapper for clang-tidy and clang static analysis.
`Codechecker analyze` creates *.plist files. These can be parsed with the already existing parsers.
In a second step Codechecker convert the contents of the *.plist files into one, human readable text file.
`Codechecker parse folder_with_plist_files > textfile.txt`.
And the new parser parses this text file.

the pull request for `warnings-ng-plugin` is
https://github.com/jenkinsci/warnings-ng-plugin/pull/1110

The following part in this message is out dated. I got it working:
Why a draft?
I was not able to make it work locally.
I build the plugins `warnings-ng` and `analysis-model-api` and uploaded the `.hpi` files to a jenkins that is not the one provided by the devenv.
The jenkins log says: it can not find "codeChecker" but lists it a few lines below anyway. ( Screen shot)
Also tried to put the new versions into the `pom.xml` of `analysis-model-api` and `warnings-ng`. No difference.
Renamed "codeChecker" to see if it affects the error message. And it does. Error message changes and the keyword in the list after the error message. 
I don't know what I am missing.
In addition to that I don't know if I got those `serialVersionUID` right. I think I just copied it from clang-tidy parser and changed one digit in the middle.

Screenshot from 2021-11-12 21-16-19
y.

### CODE REVIEW

File: src/main/java/edu/hm/hafner/analysis/parser/CodeCheckerParser.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+ * Codechecker parses the *.plist files and converts it into plain text file.
+ *
+ * Better for human readers and for using grep and diff
+ * It also puts the human-readable Severity at the start of a line.
+ *
+ */
+public class CodeCheckerParser extends LookaheadParser {
+    private static final long serialVersionUID = -3015592762345283582L;
+    private static final String CODE_CHECKER_DEFECT_PATTERN =
+            "^\\[(CRITICAL|HIGH|MEDIUM|LOW)\\] (.+):(\\d+):(\\d+): (.*?) \\[([^\\s]*?)\\]$";
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/CodeCheckerParser.java:25]: Can you please use named groups? Example: https://github.com/jenkinsci/analysis-model/blob/master/src/main/java/edu/hm/hafner/analysis/parser/PyLintParser.java

Then the code is more readable.

File: src/main/java/edu/hm/hafner/analysis/registry/CodeCheckerDescriptor.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import edu.hm.hafner.analysis.IssueParser;
+import edu.hm.hafner.analysis.parser.CodeCheckerParser;
+
+/**
+ * A descriptor for the Codechecker parser.
+ *
+ */
+class CodeCheckerDescriptor extends ParserDescriptor {
+    private static final String ID = "CodeChecker";
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/registry/CodeCheckerDescriptor.java:11]: Typically IDs are all lowercase and use a dash as separator. They are used in URLs later on.
```suggestion
    private static final String ID = "code-checker";
```

File: src/main/java/edu/hm/hafner/analysis/parser/CodeCheckerParser.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        else {
+            priority = Severity.WARNING_LOW;
+        }
+        return builder.setFileName(matcher.group(2))
+                .setSeverity(priority)
+                .setLineStart(matcher.group(3))
+                .setColumnStart(matcher.group(4))
+                .setCategory(matcher.group(6))
+                .setMessage(matcher.group(5))
+                .setType("Warning")
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/CodeCheckerParser.java:56]: No need to define a type if there is just one thing.
```suggestion
```

File: src/test/java/edu/hm/hafner/analysis/parser/CodeCheckerParserWithLinuxPathsTest.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        super("CodeChecker_with_linux_paths.txt");
+    }
+
+    @Override
+    protected CodeCheckerParser createParser() {
+        return new CodeCheckerParser();
+    }
+
+    @Override
+    protected void assertThatIssuesArePresent(final Report annotation, final SoftAssertions softly) {
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/CodeCheckerParserWithLinuxPathsTest.java:23]: Either use `report` or `annotations` (plural).
```suggestion
    protected void assertThatIssuesArePresent(final Report report, final SoftAssertions softly) {
```

File: src/test/java/edu/hm/hafner/analysis/parser/CodeCheckerParserWithWindowsPathsTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
+package edu.hm.hafner.analysis.parser;
+
+import edu.hm.hafner.analysis.AbstractParserTest;
+import edu.hm.hafner.analysis.Report;
+import edu.hm.hafner.analysis.Severity;
+import edu.hm.hafner.analysis.assertions.SoftAssertions;
+
+import static edu.hm.hafner.analysis.assertions.Assertions.*;
+
+class CodeCheckerParserWithWindowsPathsTest extends AbstractParserTest {
```
Comment (uhafner) [inline src/test/java/edu/hm/hafner/analysis/parser/CodeCheckerParserWithWindowsPathsTest.java:10]: Please combine this test case with the other one in a single file. You can add an additional test method. 
Example: https://github.com/jenkinsci/analysis-model/blob/7805d15423612fe68494f63a89d4bce6a96ce514/src/test/java/edu/hm/hafner/analysis/parser/PylintParserTest.java#L255

**Review by uhafner — CHANGES_REQUESTED** [review]
Thanks for providing a pull request!

Just some minor modifications required...

**Comment by uhafner** [comment]
Can you please also rebase with `master`? Then we can create a following pull request for the analysis-model-api plugin that uses the created (incremental) version. Example: https://github.com/jenkinsci/analysis-model-api-plugin/pull/53

**Comment by uhafner** [comment]
You will find the deployed draft release at https://repo.jenkins-ci.org/incrementals/edu/hm/hafner/analysis-model/10.7.0-rc4434.5af5868d8aee/

You can use this version in the analysis-model-api-plugin.

(You can click on the last link in the checks tab here in GitHub to find the latest version with your changes: continuous-integration/jenkins/incrementals — Deployed to Incrementals.)

**Comment by ferkulat** [comment]
@uhafner Thank you for the link.
If I understand correctly, on my jenkins instance I could use the latest `warnings-ng-plugin` release (9.7.0) and only update the `analysis-model-api-plugin`.
I uploaded a snapshot  `10.6.0-SNAPSHOT (private-449dd15e-marcel)`, where I checked that the git hash matches my commit in `analysis-model-api`. I changed this in `analysis-model-api-plugin/pom.xml`
~~~~
<analysis-model.version>10.7.0-rc4434.5af5868d8aee</analysis-model.version>
~~~~
It still does not know anything about codeChecker.
In this case ( warnings-ng-plugin not changed at all) `codeChecker` does not even get listed in known keywords. (screen shot)

Are you sure `warnings-ng-plugin` does not need to get changed?

Screenshot from 2021-11-14 17-23-22

**Comment by uhafner** [comment]
Yes this does not work yet, you get a pipeline symbol only by changing the warnings plugin as well. 

But you can use: 
```
recordIssues tool:analysisParser('code-checker') ...
```
This should work out of the box without changing the warnings plugin.

**Comment by ferkulat** [comment]
@uhafner 
Sorry for asking, but I tried the new notation in different variations.
How should I translate following to the new notation?
~~~~
recordIssues tool:codeChecker(pattern: 'b/static_analysis_report.txt', name: 'Codechecker')
~~~~

**Comment by ferkulat** [comment]
I got it working by making changes in `warnings-ng-plugin`
~~~~
recordIssues tool:codeChecker( pattern: 'b/static_analysis_report.txt', name: 'Codechecker')
~~~~
I  am going to make a pull request for `warnings-ng-plugin` as well.

But I still would like to know how to translate the Jenkinsfile line, into the alternative notation. ;)

Screenshot from 2021-11-14 19-04-16

File: src/main/java/edu/hm/hafner/analysis/parser/CodeCheckerParser.java [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        }
+        else if (matcher.group("severity").contains("HIGH")) {
+            priority = Severity.WARNING_HIGH;
+        }
+        else if (matcher.group("severity").contains("MEDIUM")) {
+            priority = Severity.WARNING_NORMAL;
+        }
+        else {
+            priority = Severity.WARNING_LOW;
+        }
```
Comment (uhafner) [inline src/main/java/edu/hm/hafner/analysis/parser/CodeCheckerParser.java:49]: Sorry for being pedantic: can you move that part to a method and extract the duplication `matcher.group("severity")` into a variable. At least the latter code smell would be nice to fix

**Review by uhafner — APPROVED** [review]
One final call of https://github.com/jenkinsci/analysis-model/blob/master/src/main/java/edu/hm/hafner/analysis/registry/ParserRegistry.java#L229 would be helpful to get the new parser listed automatically.

File: src/main/java/edu/hm/hafner/analysis/parser/CodeCheckerParser.java (hunk 6 above)
Comment (ferkulat) [reply src/main/java/edu/hm/hafner/analysis/parser/CodeCheckerParser.java:49]: When extracting into `getSeverity` I would like to go one step further, and remove state.
But I would like to know if that would match your taste:
~~~~
    private Severity getSeverity(final String severityText) {
        if (severityText.contains("CRITICAL")) { return Severity.ERROR;         }
        if (severityText.contains("HIGH"    )) { return Severity.WARNING_HIGH;  }
        if (severityText.contains("MEDIUM"  )) { return Severity.WARNING_NORMAL;}

        return  Severity.WARNING_LOW;
    }
~~~~

Comment (uhafner) [reply src/main/java/edu/hm/hafner/analysis/parser/CodeCheckerParser.java:49]: Yes that would be even simpler👍 (I think CheckStyle will complain that the {} are in one line though.)

**Comment by elupus** [comment]
CodeChecker can also output severity of UNSPECIFIED which are now missed by this plugin.

**Pull request merged**